l10n_ch_base_bank/models/invoice.py

```diff
@@ -121,15 +117,39 @@ def _check_isr(self):
                 invoice._is_isr_reference()
         return True
 
+    def write(self, vals):
+        """Override to update partner_bank_id before constraints if needed and
+        to be consistent with create
+        """
+        res = super().write(vals)
+        for rec in self:
+            if not rec.partner_bank_id and rec.type == 'out_invoice':
+                banks = rec.partner_banks_to_show()
+                if banks:
+                    rec.write({'partner_bank_id': banks[0].id})
+        return res
+
     @api.model
     def create(self, vals):
-        """We override create in order to have customer invoices
-        generated by the comercial flow as on change partner is
-        not systemtically call"""
+        # make sure to have customer invoices generated by the commercial flow
+        # since on partner change it doesn't happen automatically
         type_defined = vals.get('type') or self.env.context.get('type', False)
         if type_defined == 'out_invoice' and not vals.get('partner_bank_id'):
-            user = self.env.user
-            bank_ids = user.company_id.partner_id.bank_ids
-            if bank_ids:
-                vals['partner_bank_id'] = bank_ids[0].id
+            partner = self.env.user.company_id.partner_id
+            vals['partner_bank_id'] = self._get_bank_id(
+                partner, vals.get('journal_id'), vals.get('reference_type'),
+            )
```

Comment:
In which case can't we rely on `partner_banks_to_show`?

Can't we let `partner_bank_id` undefined unless it is manually set by the user?

One reason we could want to copy the bank on invoice is for the user to see it.
Nevertheless, I would argue that the payment mode should already allow us to invoice partners by multiple means.
And that setting the bank on the invoice should be an exception and can be done manually.

I fear that we still are making something different than the OCA standard for other localisations. I would prefer to keep the partner_bank_id empty and only refer to `partner_banks_to_show`.

We probably want to find the payment mode instead.

Reply:
this changes we made to satisfy this constrain https://github.com/OCA/l10n-switzerland/blob/37e940e6977ef0fc63b1fce526173a4cd5cd36f5/l10n_ch_base_bank/models/invoice.py#L72
partner_banks_to_show will work on already created obj

Reply:
if we dropped constraint https://github.com/OCA/l10n-switzerland/blob/37e940e6977ef0fc63b1fce526173a4cd5cd36f5/l10n_ch_base_bank/models/invoice.py#L86
we can rid of create overriding, than only one test is failed https://github.com/OCA/l10n-switzerland/blob/37e940e6977ef0fc63b1fce526173a4cd5cd36f5/l10n_ch_base_bank/tests/test_create_invoice.py#L81
@yvaucher  maybe that test and behavior is obsolete?

Reply:
@yvaucher ping

Reply:
How I see it:

1. As a user, I create an invoice.
2. I set the partner. (at this point the payment mode can be set by the onchange [1] )
3. I set the reference type to ISR
4. Then the constraint check if bank account (either on invoice or in payment mode) is compatible with ISR type.
   if yes -> nothing
   else -> raise error

In case of error, the user will have to change manually the payment mode or the partner_bank_id. (The error message needs to be adapted accordingly to explain that payment mode or invoice bank)

About `_get_bank_id`, it could be added to:
- the partner onchange of OCA/bank-payment/account_payment_partner
- plus a reference_type onchange.
But only if the reference type is ISR and it should also look at the payment modes.
Thus, I propose to drop it for the current PR and if the needs come we will add it in another PR.


TODO:

- Thus we can drop the part where we copy the bank account on the invoice `write` and `create`. I prefer to let the module `account_payment_partner`'s onchange do the job.
- Change the constraint to check with `partner_banks_to_show`
- Change the constraint message to reflect the possible use of either payment mode or bank field on invoice.
- Adapt the test you mentioned by calling the partner onchange or by setting the partner_bank_id (the check itself is about the constraint which is still valid)
- Drop `_get_bank_id` which could come in a further PR.


Reasons:

1. IMO we should stick to OCA/bank-payment behavior as much as possible as you could have a multi-company setup and you don't want the swiss company to mess with other country companies.
2. Less code make me more happy as long as we make it simple but not simpler :)


[1]
https://github.com/OCA/bank-payment/blob/1a21770d76ec3af9fa918b8256628257884113e9/account_payment_partner/models/account_invoice.py#L47-L49